pilot/pkg/model/jwks_resolver.go

```diff
@@ -207,39 +220,36 @@ func newJwksResolverWithCABundlePaths(evictionDuration, refreshInterval, retryIn
 	return ret
 }
 
-// ResolveJwksURI sets jwks_uri through openID discovery if it's not set in request authentication policy.
-func (r *JwksResolver) ResolveJwksURI(policy *v1beta1.RequestAuthentication) {
-	for _, rule := range policy.JwtRules {
-		if rule.JwksUri == "" && rule.Jwks == "" {
-			if uri, err := r.resolveJwksURIUsingOpenID(rule.Issuer); err == nil {
-				rule.JwksUri = uri
-			} else {
-				log.Warnf("Failed to get jwks_uri for issuer %q: %v", rule.Issuer, err)
-			}
-		}
-	}
-}
-
 // GetPublicKey gets JWT public key and cache the key for future use.
-func (r *JwksResolver) GetPublicKey(jwksURI string) (string, error) {
+func (r *JwksResolver) GetPublicKey(issuer string, jwksURI string) (string, error) {
 	now := time.Now()
-	if val, found := r.keyEntries.Load(jwksURI); found {
+	key := jwtKey{issuer: issuer, jwksURI: jwksURI}
+	if val, found := r.keyEntries.Load(key); found {
 		e := val.(jwtPubKeyEntry)
 		// Update cached key's last used time.
 		e.lastUsedTime = now
-		r.keyEntries.Store(jwksURI, e)
+		r.keyEntries.Store(key, e)
 		return e.pubKey, nil
 	}
 
-	// Fetch key if it's not cached.
+	if jwksURI == "" {
+		// Fetch the jwks URI if it is not hardcoded on config.
+		var err error
+		jwksURI, err = r.resolveJwksURIUsingOpenID(issuer)
+		if err != nil {
+			log.Errorf("Failed to jwks URI from %q: %v", issuer, err)
+			return "", err
+		}
+	}
+
 	resp, err := r.getRemoteContentWithRetry(jwksURI, networkFetchRetryCountOnMainFlow)
```

Comment:
I realize this is not made worse by your PR, but I am actually pretty concerned about sending these network calls (with retry) on the XDS generation. Not trying to block this PR, but we may need to investigate this. I have concerns about how it will behave at scale

Reply:
As the `networkFetchRetryCountOnMainFlow` is set to 0 anyway we could also just remove the immediate retries and rely on the refresh mechanism. The refreshs are usually happening every 20 minutes and with the change of this PR on error they are happening 1m, 2m, 4m, 8m, 16m, 20m, 20m...

Reply:
the current way is to block the xDS a little bit for the network request (the fetched jwks will be cached) and then push the jwks within the same single xDS response, the same result will be cached and reused for all pods, how is this related to the scale of pods?

The other way is to let the separate refresh goroutine to fetch and push asynchronously, but that means an extra xDS push just for the jwks key, and until it completes the request authentication will have no effect.